primitives/npos-elections/src/phragmms.rs

```diff
@@ -0,0 +1,380 @@
+///
+/// - The resulting edge weight distribution is normalized (thus, safe to use for submission).
+/// - The accuracy can be configured via the generic type `P`.
+///
+/// This can only fail of the normalization fails.
+pub fn phragmms<AccountId: IdentifierT, P: PerThing>(
+	to_elect: usize,
+	initial_candidates: Vec<AccountId>,
+	initial_voters: Vec<(AccountId, VoteWeight, Vec<AccountId>)>,
+) -> Result<ElectionResult<AccountId, P>, &'static str>
+	where ExtendedBalance: From<InnerOf<P>>
+{
+	let (candidates, mut voters) = setup_inputs(initial_candidates, initial_voters);
+
+	let mut winners = vec![];
+	for round in 0..to_elect {
+		let round_winner = calculate_max_score::<AccountId, P>(&candidates, &voters);
+		apply_elected::<AccountId>(&mut voters, Rc::clone(&round_winner));
+
+		round_winner.borrow_mut().round = round;
+		round_winner.borrow_mut().elected = true;
+		winners.push(round_winner);
+
+		balance(&mut voters, 2, 0);
+	}
+
+	let mut assignments = voters.into_iter().filter_map(|v| v.into_assignment()).collect::<Vec<_>>();
+	let _ = assignments.iter_mut().map(|a| a.try_normalize()).collect::<Result<(), _>>()?;
+	let winners = winners.into_iter().map(|w_ptr|
+		(w_ptr.borrow().who.clone(), w_ptr.borrow().backed_stake)
+	).collect();
+
+	Ok(ElectionResult { winners, assignments })
+}
+
+/// Find the candidate that can yield the maximum score for this round.
+///
+/// Returns a new `CandidatePtr` to the winner candidate. The score of the candidate is updated and
+/// can be read from the returned pointer.
+///
+/// This is an internal part of the [`balanced_heuristic`].
+pub(crate) fn calculate_max_score<AccountId: IdentifierT, P: PerThing>(
+	candidates: &[CandidatePtr<AccountId>],
+	voters: &[Voter<AccountId>],
+) -> CandidatePtr<AccountId> where ExtendedBalance: From<InnerOf<P>> {
+	for c_ptr in candidates.iter() {
+		let mut candidate = c_ptr.borrow_mut();
+		if !candidate.elected {
+			candidate.score = Rational128::from(1, P::ACCURACY.into());
+		}
+	}
+
+	for voter in voters.iter() {
+		let mut denominator_contribution: ExtendedBalance = 0;
+
+		// gather contribution from all elected edges.
+		for edge in voter.edges.iter() {
+			let edge_candidate = edge.candidate.borrow();
+			if edge_candidate.elected {
+				let edge_contribution: ExtendedBalance = P::from_rational_approximation(
+					edge.weight,
+					edge_candidate.backed_stake,
+				).deconstruct().into();
+				denominator_contribution += edge_contribution;
+			}
+		}
+
+		// distribute to all _unelected_ edges.
+		for edge in voter.edges.iter() {
+			let mut edge_candidate = edge.candidate.borrow_mut();
+			if !edge_candidate.elected {
+				let prev_d = edge_candidate.score.d();
+				edge_candidate.score = Rational128::from(1, denominator_contribution + prev_d);
+			}
+		}
+	}
+
+	// finalise the score value, and find the best.
+	let mut best_score = Rational128::zero();
+	let mut best_candidate = Rc::clone(&candidates[0]);
```

Comment:
what if candidates[0] is already elected ? and there is no longer unelected candidates ?

More globally, how phragmms is suppose to behave if `to_elect > candidates.len()`

Reply:
This is indeed a bug, albeit it will probably never be triggered because we will always find a better candidate with a better `best_score`, thus candidate[0] is just a placeholder. 

> More globally, how phragmms is suppose to behave if to_elect > candidates.len()

All of the implementation are best-effort, meaning that the user is responsible for providing enough candidates and the algorithm _tries_ to return `to_elect`. 

Previously, seq-phragmen had a `min_to_elect` and if it was provided with less candidates, it returned `None`. But I changed this now because I think it is a redundant API. the call site of these election algorithms can easily ensure that they are passing in enough candidates.

Reply:
This needs fixing: currently, even if the behaviour is correct, the implementation is blind to the fact that we can `to_elect > candidates.len()` and at some point it should just stop. This can easily be done when we iterate over all candidates in each round, and if we realise that all of them are `elected = true`.

Reply:
Yes I acknowledge Kusama and other big chains will have enough candidates, but some tests or unreliable or buggy chains might end up having not enough candidates and in this case behavior seems to be very unexpected. I guess its fine, maybe log some error or at least document the assumption.

Reply:
This is fixed: the function now returns `Option<_>` and I have added this to the documentation. 

Note that the documentation of this function is a bit lazy and says: _has the same API as `seq-phragmen`_. If you see `seq-phragmen`, this is explained in greater detail.